source/opt/split_combined_image_sampler_pass.cpp

```diff
@@ -0,0 +1,611 @@
+      if (inst.opcode() == spv::Op::OpTypeFunction) {
+        analysis::Function* f_ty =
+            type_mgr_->GetType(inst.result_id())->AsFunction();
+        std::vector<const analysis::Type*> new_params;
+        for (const auto* param_ty : f_ty->param_types()) {
+          const auto param_ty_id = type_mgr_->GetId(param_ty);
+          if (combined_types_.find(param_ty_id) != combined_types_.end()) {
+            auto* param_type = def_use_mgr_->GetDef(param_ty_id);
+            auto [image_type, sampler_type] = SplitType(*param_type);
+            assert(image_type);
+            assert(sampler_type);
+            // The image and sampler types must already exist, so there is no
+            // need to move them to the right spot.
+            new_params.push_back(type_mgr_->GetType(image_type->result_id()));
+            new_params.push_back(type_mgr_->GetType(sampler_type->result_id()));
+          } else {
+            new_params.push_back(param_ty);
+          }
+        }
+        if (new_params.size() != f_ty->param_types().size()) {
+          // Replace this type.
+          analysis::Function new_f_ty(f_ty->return_type(), new_params);
+          const uint32_t new_f_ty_id = type_mgr_->GetTypeInstruction(&new_f_ty);
+          def_use_mgr_->ForEachUse(&inst,
+                                   [&](Instruction* user, uint32_t use_index) {
+                                     user->SetOperand(use_index, {new_f_ty_id});
+                                     reanalyze_set.insert(user);
+                                   });
+          MarkAsDead(&inst);
+
+          reanalyze_set.insert(def_use_mgr_->GetDef(new_f_ty_id));
+          // Reanalyze the non-combined parameter types, and the return type.
+          auto* new_f_ty_inst = def_use_mgr_->GetDef(new_f_ty_id);
+          new_f_ty_inst->ForEachId([&](const uint32_t* param_id_ptr) {
+            reanalyze_set.insert(def_use_mgr_->GetDef(*param_id_ptr));
+          });
+        }
+      }
+    }
+    for (auto* inst : reanalyze_set) {
+      def_use_mgr_->AnalyzeInstDefUse(inst);
+    }
+  }
+
+  // Rewite OpFunctionParameter in function definitions.
+  for (Function& fn : *context()->module()) {
+    std::vector<Instruction*> to_replace;
+    fn.ForEachParam([&](Instruction* param) {
+      auto param_ty_id = param->type_id();
+      if (combined_types_.find(param_ty_id) != combined_types_.end()) {
+        to_replace.push_back(param);
+      }
+    });
+    if (to_replace.empty()) {
+      continue;
+    }
+    auto next_to_replace = to_replace.begin();
+    struct Replacement {
+      Instruction* combined;
+      Instruction* image;
+      Instruction* sampler;
+    };
+    std::vector<Replacement> replacements;
+    Function::RewriteParamFn rewriter =
+        [&](std::unique_ptr<Instruction>&& from_param,
+            std::back_insert_iterator<Function::ParamList>& appender) {
+          auto param = std::move(from_param);
+          if (param.get() == *next_to_replace) {
```

Comment:
nit: flip condition, early-return & denest.

Reply:
i don't like the rearly return and de-nest in this case